receiver/prometheusexecreceiver/subprocessmanager/manager.go

```diff
@@ -80,23 +64,40 @@ func (proc *Process) Run(logger *zap.Logger) (time.Duration, error) {
 	go proc.pipeSubprocessOutput(bufio.NewReader(stderrReader), logger, false)
 
 	// Start and stop timer (elapsed) right before and after executing the command
+	done := make(chan error, 1)
 	start := time.Now()
+
 	errProcess := childProcess.Start()
 	if errProcess != nil {
 		return 0, fmt.Errorf("process could not start: %w", errProcess)
 	}
 
-	errProcess = childProcess.Wait()
-	elapsed := time.Since(start)
-	if errProcess != nil {
-		return elapsed, fmt.Errorf("process error: %w", errProcess)
-	}
+	go func() {
+		done <- childProcess.Wait()
+	}()
+
+	// Handle normal process exiting or parent logic triggering a shutdown
+	select {
+	case errProcess := <-done:
+		elapsed := time.Since(start)
 
-	return elapsed, nil
+		// The subprocess will most likely exit due to an error (which is logged), but in any case we'll continue as usual with the health/delay computation
+		if errProcess != nil {
+			return elapsed, fmt.Errorf("process error: %w", errProcess)
+		}
+		return elapsed, nil
+
+	case <-ctx.Done():
+		err := childProcess.Process.Kill()
+		if err != nil {
+			logger.Error("couldn't kill subprocess", zap.String("error", err.Error()))
+		}
+		return 0, nil
```

Comment:
This case will only trigger if the receiver is shutting down, which means this return will be ignored since the `select` in the receiver has already had a case triggered, and the receiver loop will be exited, also meaning that elapsed's value is useless!

Reply:
That's also why I log the error there and don't return it, since it'll be ignored.

Reply:
Yea but my point is you can't assume what the client code (the receiver) is doing here. This code could be called by another package that might expect the duration to be returned properly

Reply:
Perfect, understood